src/core/FObject.js

```diff
@@ -169,36 +185,70 @@ foam.CLASS({
       while ( listeners ) {
         var l = listeners.l;
         var s = listeners.sub;
-        // Like l.apply(l, [s].concat(Array.from(a))), but faster.
-        // FUTURE: add benchmark to justify
-        // ???: optional exception trapping, benchmark
-        switch ( a.length ) {
-          case 0: l(s); break;
-          case 1: l(s, a[0]); break;
-          case 2: l(s, a[0], a[1]); break;
-          case 3: l(s, a[0], a[1], a[2]); break;
-          case 4: l(s, a[0], a[1], a[2], a[3]); break;
-          case 5: l(s, a[0], a[1], a[2], a[3], a[4]); break;
-          case 6: l(s, a[0], a[1], a[2], a[3], a[4], a[5]); break;
-          case 7: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6]); break;
-          case 8: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]); break;
-          case 9: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8]); break;
-          default: l.apply(l, [s].concat(Array.from(a)));
-        }
+        this.notifyListener_(l, s, a);
         listeners = listeners.next;
         count++;
       }
       return count;
     },
 
+    function notifyListener_(l, s, a) {
+      // Like l.apply(l, [s].concat(Array.from(a))), but faster.
+      // FUTURE: add benchmark to justify
+      // ???: optional exception trapping, benchmark
+      switch ( a.length ) {
+      case 0: l(s); break;
+      case 1: l(s, a[0]); break;
+      case 2: l(s, a[0], a[1]); break;
+      case 3: l(s, a[0], a[1], a[2]); break;
+      case 4: l(s, a[0], a[1], a[2], a[3]); break;
+      case 5: l(s, a[0], a[1], a[2], a[3], a[4]); break;
+      case 6: l(s, a[0], a[1], a[2], a[3], a[4], a[5]); break;
+      case 7: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6]); break;
+      case 8: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]); break;
+      case 9: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8]); break;
+      default: l.apply(l, [s].concat(Array.from(a)));
+      }
+    },
+
     function hasListeners(/* args */) {
       /** Return true iff there are listeners for the supplied message. **/
+      var args = Array.from(arguments);
+      return this.hasKeyedListeners_(args) || this.hasExprListeners_(args);
```

Comment:
Isn't this inefficient because both methods walk the same list?  Could they be combined?